sirepo/pkcli/runner.py

```diff
@@ -105,122 +93,192 @@ def _write_status(status, run_dir):
+            return
+        pkdlog(
+            'kill_all {}: killing job with jhash {}',
+            run_dir, job_info.jhash,
+        )
+        job_info.cancel_requested = True
+        await job_info.report_job.kill(_KILL_TIMEOUT_SECS)
+
+    async def start_report_job(self, run_dir, jhash, backend, cmd, tmp_dir):
+        # First make sure there's no-one else using the run_dir
+        current_jhash, current_status = self.run_dir_status(run_dir)
+        if current_status is runner_client.JobStatus.RUNNING:
+            # Something's running.
+            if current_jhash == jhash:
+                # It's already the requested job, so we have nothing to
+                # do. Throw away the tmp_dir and move on.
+                pkdlog(
+                    '{} {}: job is already running; skipping',
+                    run_dir, jhash,
+                )
+                pkio.unchecked_remove(tmp_dir)
                 return
-            try:
-                env = _subprocess_env()
-                run_log_path = run_dir.join(template_common.RUN_LOG)
-                # we're in py3 mode, and regular subprocesses will inherit our
-                # environment, so we have to manually switch back to py2 mode.
-                env['PYENV_VERSION'] = 'py2'
-                cmd = ['pyenv', 'exec'] + cmd
-                with open(run_log_path, 'a+b') as run_log:
-                    process = trio.Process(
-                        cmd,
-                        cwd=run_dir,
-                        start_new_session=True,
-                        stdin=subprocess.DEVNULL,
-                        stdout=run_log,
-                        stderr=run_log,
-                        env=env,
-                    )
-                self.jobs[run_dir] = _JobInfo(
-                    run_dir, jhash, runner_client.JobStatus.RUNNING, process
+            else:
+                # It's some other job. Better kill it before doing
+                # anything else.
+                # XX TODO: should we check some kind of sequence number
+                # here? I don't know how those work.
+                pkdlog(
+                    '{} {}: stale job is still running; killing it',
+                    run_dir, jhash,
                 )
-                async with process:
-                    task_status.started()
-                    # XX more race conditions here, in case we're writing to
-                    # the wrong version of the directory...
-                    await process.wait()
-                    if process.returncode:
+                await self.kill_all(run_dir)
+
+        # Okay, now we have the dir to ourselves. Set up the new run_dir:
+        assert run_dir not in self.report_jobs
+        pkio.unchecked_remove(run_dir)
+        tmp_dir.rename(run_dir)
+        # Start the job:
+        report_job = await _BACKENDS[backend].start_report_job(run_dir, cmd)
+        # And update our records so we know it's running:
```

Comment:
The locking strategy is: each `run_dir` has an associated lock. Everything that touches the `run_dir`-related data structures holds that lock. Currently this is done mostly by `handle_conn` (it takes the lock before calling the handler), and then `_supervise_report_job` also takes the lock after the job finishes so it can write back the new state.

I'm not very happy about how implicit this is in the code, and have been trying to figure out if there's a way to make it more clear. (Maybe using reentrant locks and re-acquiring the lock in every method?)